jsonschema/validators.py

```diff
@@ -356,7 +358,7 @@ def __init__(
         self._remote_cache = remote_cache
 
     @classmethod
-    def from_schema(cls, schema, *args, **kwargs):
+    def from_schema(cls, schema, scope_key="id", *args, **kwargs):
```

Comment:
So, this also makes me uneasy -- up until now, ref resolvers were completely separate from drafts (because the behavior was the same across them).

I don't love this idea -- now a caller needs to know implementation details of the draft they're using.

I'd much rather encapsulate that, even if that means that we now need to have separate ref resolvers per draft (and we'd also need to preserve exactly the old interface for a class called just `RefResolver`). Annoying, but I think the whole change is annoying :/

Reply:
> So, this also makes me uneasy -- up until now, ref resolvers were completely separate from drafts (because the behavior was the same across them).

The problem is that they're not, because that `from_schema` method needs to understand id. My initial thought was actually to drop that method completely, but of course that's extremely unpleasant for users in the wild.

Reply:
Right but what I mean is -- as far as `RefResolver` is concerned, when presented with any schema, there is an algorithm you can use to unambiguously know this (but now this introduces the possibility of error).

Even now, you have this algorithm:

* Look for a `$schema` key
* If you find one, you (RefResolver) know what the correct scope key is
* If not, you use the "package" default (which I'm not too happy with how we're managing currently :/, but it's a thing)

You can take that algorithm, and layer it on top of an "explicit" one -- say a `Draft3Draft4RefResolver` and a `Draft6RefResolver` alongside a `resolver_for_schema` that calls the right thing as above.

(I know I'm being terse still, lemme know if that's unclear)